Objects/iterobject.c

```diff
@@ -288,3 +288,323 @@ PyTypeObject PyCallIter_Type = {
 };
 
 
+/* -------------------------------------- */
+
+typedef struct {
+    PyObject_HEAD
+    PyObject *it_callable; /* Set to NULL when iterator is exhausted */
+    PyObject *it_sentinel; /* Set to NULL when iterator is exhausted */
+} callasynciterobject;
+
+typedef struct {
+    PyObject_HEAD
+    PyObject *wrapped_iter; /*
+        The iterator returned by the callable, unwrapped via __await__.
+
+        If NULL this means that the iterator is already exhausted and when
+        iterated it should raise StopAsyncIteration.
+    */
+    callasynciterobject *it; /* The iterator object, in order to clear it when done. */
+} callasynciterawaitableobject;
+
+
+PyObject *
+PyCallAsyncIter_New(PyObject *callable, PyObject *sentinel)
+{
+    callasynciterobject *it;
+    it = PyObject_GC_New(callasynciterobject, &PyCallAsyncIter_Type);
+    if (it == NULL)
+        return NULL;
+    Py_INCREF(callable);
+    it->it_callable = callable;
+    Py_INCREF(sentinel);
+    it->it_sentinel = sentinel;
+    _PyObject_GC_TRACK(it);
+    return (PyObject *)it;
+}
+
+static void
+callasynciter_dealloc(callasynciterobject *obj)
+{
+    _PyObject_GC_UNTRACK(obj);
+    Py_XDECREF(obj->it_callable);
+    Py_XDECREF(obj->it_sentinel);
+    PyObject_GC_Del(obj);
+}
+
+static void
+callasynciterawaitable_dealloc(callasynciterawaitableobject *obj)
+{
+    _PyObject_GC_UNTRACK(obj);
+    Py_XDECREF(obj->wrapped_iter);
+    Py_XDECREF(obj->it);
+    PyObject_GC_Del(obj);
+}
+
+static int
+callasynciter_traverse(callasynciterobject *it, visitproc visit, void *arg)
+{
+    Py_VISIT(it->it_callable);
+    Py_VISIT(it->it_sentinel);
+    return 0;
+}
+
+static int
+callasyncawaitable_traverse(callasynciterawaitableobject *obj, visitproc visit, void *arg)
+{
+    Py_VISIT(obj->wrapped_iter);
+    Py_VISIT(obj->it);
+    return 0;
+}
+
+static PyObject *
+callasynciter_anext(callasynciterobject *it)
+{
+    PyObject *obj = NULL, *wrapped_iter = NULL;
+    PyTypeObject *t = NULL;
+    callasynciterawaitableobject *awaitable;
+
+    if (it->it_callable == NULL) {
+        /* Can we raise this at this point, or do we need to return an awaitable
+         * that raises it? */
+        PyErr_SetNone(PyExc_StopAsyncIteration);
```

Comment:
I assume this comment is for reviewers, and should be removed / addressed before merging? If so, leaving this comment as a marker for someone who knows more.

Reply:
Removed this comment in the latest revision so this PR is in a more mergeable state, but leaving this conversation unresolved until we get confirmation. The removed comment was:
> Can we raise this at this point, or do we need to return an awaitable that raises it?

...and referred to (what's currently on line 368), `PyErr_SetNone(PyExc_StopAsyncIteration);`.